File: C/quicksort.c
#include <stdio.h>
int main(int argc, char const *argv[])
{
    int arr[] = {2, 9, 3, 7, 4};
    int n = 5;
    // insertionsort(arr,n);
    print(arr,n);
    quicksort(arr, 0, n-1);
    print(arr,n);
    return 0;
}
int partition(int *arr, int low, int high){
int i,j,temp;
do{
    int pivot = arr[low];
    i = low + 1;
     j = high;
    while (arr[i] <= pivot)
    {
        i++;
    }
    while (arr[i] > pivot)
    {
        j--;
    }
    if (i < j)
    {
        temp = arr[low];
        arr[low] = arr[j];
        arr[j] = temp;
    }
}while(i<j);
}
int print(int *arr, int n)
{
    for (int i = 0; i < n; i++)
    {
        printf("%d ", arr[i]);
    }
}
int quicksort(int *arr, int low, int high)
{
    int pi; //index of pivot after partition
    if (low < high)
    {
        pi = partition(arr, low, high);
        quicksort(arr, low, pi - 1); //sort left subarray
        quicksort(arr, pi + 1, high); //sort right subarray
    }
}

File: C#/QuickSort.cs
using System;
					
public class Program
{
	public static void Main()
	{
		int[] arr = new int[] { -1, -5, 2, 10, 0, 4, 9 };
		foreach (int a in arr) Console.Write(a + " ");
		Console.WriteLine();
		
		QuickSort quick = new QuickSort();
		quick.Sort(arr, 0, 6);
		foreach (int a in arr) Console.Write(a + " ");
	}
}

public class QuickSort
{
	public int Partition(int[] arr, int left, int right)
	{
		int pivot; int temp;
		pivot = arr[left];
		while (true)
		{
			while (arr[left] < pivot)
			{
				left++;
			}
			while (arr[right] > pivot)
			{
				right--;
			}
			if (left < right)
			{
				temp = arr[right];
				arr[right] = arr[left];
				arr[left] = temp;
			}
			else
			{
				return right;	
			}
		}
	}
	
	public void Sort(int[] arr, int left, int right)
	{
		int pivot;
		if (left < right)
		{
			pivot = Partition(arr, left, right);
			if (pivot > 1)
			{
				Sort(arr, left, pivot - 1);
			}
			if (pivot + 1 < right)
			{
				Sort(arr, pivot + 1, right);
			}
		}
	}
}
